pkg/catalogv2/git/download.go

```diff
@@ -1,105 +1,133 @@
+	staticDir := filepath.Join(staticDir, namespace, name, hash(gitURL))
+	if s, err := os.Stat(staticDir); err == nil && s.IsDir() {
+		return staticDir
 	}
-
-	// Try to reset to the given branch, if success exit
-	localBranchFullName := plumbing.NewBranchReferenceName(branch)
-	err = r.hardReset(localBranchFullName.String())
-	if err == nil {
-		return nil
+	localDir := filepath.Join(localDir, namespace, name, hash(gitURL))
+	if s, err := os.Stat(localDir); err == nil && s.IsDir() {
+		return localDir
 	}
+	return filepath.Join(stateDir, namespace, name, hash(gitURL))
+}
 
-	// If we do not have the branch locally, fetch and reset
-	err = r.fetchAndReset(branch)
+func Head(secret *corev1.Secret, namespace, name, gitURL, branch string, insecureSkipTLS bool, caBundle []byte) (string, error) {
+	git, err := gitForRepo(secret, namespace, name, gitURL, insecureSkipTLS, caBundle)
 	if err != nil {
-		return fmt.Errorf("failed to fetch and/or reset at branch: %w", err)
+		return "", err
 	}
-	return nil
+
+	return git.Head(branch)
 }
 
-// Head clones or sets a local repository to the HEAD of a git branch and return its commit hash.
-func (r *Repository) Head(branch string) (string, error) {
-	err := r.cloneOrOpen(branch)
+func Update(secret *corev1.Secret, namespace, name, gitURL, branch string, insecureSkipTLS bool, caBundle []byte) (string, error) {
+	git, err := gitForRepo(secret, namespace, name, gitURL, insecureSkipTLS, caBundle)
 	if err != nil {
-		return "", fmt.Errorf("failed to clone or open repository: %w", err)
+		return "", err
 	}
 
-	err = r.hardReset(string(plumbing.HEAD))
-	if err != nil {
-		return "", fmt.Errorf("failed to hard reset at HEAD: %w", err)
+	if isBundled(git) && settings.SystemCatalog.Get() == "bundled" {
+		return Head(secret, namespace, name, gitURL, branch, insecureSkipTLS, caBundle)
 	}
 
-	commit, err := r.getCurrentCommit()
-	if err != nil {
-		return commit.String(), fmt.Errorf("failed to get current commit: %w", err)
+	commit, err := git.Update(branch)
+	if err != nil && isBundled(git) {
+		return Head(secret, namespace, name, gitURL, branch, insecureSkipTLS, caBundle)
 	}
-	return commit.String(), nil
+	return commit, err
 }
 
-// CheckUpdate will check if rancher is in bundled mode,
-// if it is not in bundled mode, will make an update.
-// if it is in bundled mode, will just call Head method.
-func (r *Repository) CheckUpdate(branch, systemCatalogMode string) (string, error) {
-	if isBundled(r.Directory) && systemCatalogMode == "bundled" {
-		return r.Head(branch)
+func Ensure(secret *corev1.Secret, namespace, name, gitURL, commit string, insecureSkipTLS bool, caBundle []byte) error {
+	if commit == "" {
+		return nil
+	}
+	git, err := gitForRepo(secret, namespace, name, gitURL, insecureSkipTLS, caBundle)
+	if err != nil {
+		return err
 	}
 
-	commit, err := r.Update(branch)
-	if err != nil && isBundled(r.Directory) {
-		return r.Head(branch)
+	// If the repositories are rancher managed and if bundled is set
```

Comment:
Is this right? 
This should be Inside the other Ensure Method after the hard Reset to commit and before the fetch right?